assign-2/src/data/ArrayListStudentDao.java

```diff
@@ -0,0 +1,51 @@
+package data;
+
+import data.StudentDao;
+import logic.Student;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ArrayListStudentDao implements StudentDao {
+    private List<Student> students = new ArrayList<>();
+
+    @Override
+    public void add(Student student) {
+        students.add(student);
+    }
+
+    @Override
+    public void update(Student student) {
+        Student s = getById(student.getId());
+        if (s != null) {
+            s.setFirstName(student.getFirstName());
+            s.setLastName(student.getLastName());
+        }
+    }
+
+    @Override
+    public void delete(int id) {
+        Student s = getById(id);
+        if (s != null) {
+            students.remove(s);
+        } else {
+            throw new NoSuchStudentException(id);
+        }
+    }
+
+
+    @Override
+    public Student getById(int id) {
+        for (Student student : students) {
+            if (student.getId() == id) {
+                return student;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public List<Student> getAll() {
+        return students;
```

Comment:
Here you are returning a reference to the internal object `List<Student> students`. It is indeed, the list of all the stored students, but if you modify the list outside, those changes will be reflected in this object. Maybe you just want to return all students, but as a copy of the original collection (https://www.baeldung.com/java-copy-list-to-another)?

Reply:
`public List<Student> getAll() {
        List<Student> studentsCopy = new ArrayList<>(students);
        return studentsCopy;
}`
like this?